Fix ROCWMMA unsupported architecture errors

When compiling with ROCm 6.4.1 with a W6600 GPU, the following error and warnings are encountered:

```
/var/software/opt/rocm-6.4.1/include/rocwmma/internal/./layout/../config.hpp:76:15: error: static assertion failed: Unsupported architecture
   76 | static_assert(0, "Unsupported architecture");
      |               ^
/var/software/opt/rocm-6.4.1/include/rocwmma/rocwmma_impl.hpp:377:68: warning: 'exec<const HIP_vector_type<hip_bfloat16, 4> &>' is deprecated: Register layout transform is not supported [-Wdeprecated-declarations]
  377 |                     PackAcc::unpack(Mma::exec(PackA::pack(PreMmaA::exec(a.mAccess)),
      |                                                                    ^
/tmp/comgr-7a4570/input/gemm_kernel.cu:736:15: note: in instantiation of function template specialization 'rocwmma::mma_sync<16U, 16U, 16U, hip_bfloat16, hip_bfloat16, rocwmma::row_major, rocwmma::col_major, void, void>' requested here
  736 |         wmma::mma_sync(sum[REAL][m][n], a[REAL][m], b[REAL][n],
      |               ^
/var/software/opt/rocm-6.4.1/include/rocwmma/internal/./layout/register_layout_transforms.hpp:181:13: note: 'exec<const HIP_vector_type<hip_bfloat16, 4> &>' has been explicitly marked deprecated here
  181 |             ROCWMMA_UNSUPPORTED_IMPL("Register layout transform is not supported")
      |             ^
/var/software/opt/rocm-6.4.1/include/rocwmma/internal/./layout/../config.hpp:182:54: note: expanded from macro 'ROCWMMA_UNSUPPORTED_IMPL'
  182 | #define ROCWMMA_UNSUPPORTED_IMPL(MSG) __attribute__((deprecated(MSG)))
      |                                                      ^
```

Based on this, support for `bfloat16` should be evaluated during run time, throwing an error when the GPU does not support this operation.

--- a/include/ccglib/common/arch.h
+++ b/include/ccglib/common/arch.h
@@ -36,18 +36,18 @@ bool isVolta(cu::Device &device) {
 
 bool isUnsupported(cu::Device &device) {
   const std::string arch(device.getArch());
-  // AMD: unsupported are gfx < 8, Vega gfx9, RDNA2=gfx10
+  // AMD: only architectures with matrix cores are supported: CDNA or newer,
+  // RDNA3 or newer
 #if defined(__HIP_PLATFORM_AMD__)
-  if (arch.find("gfx6") != std::string::npos)
+  const std::vector<std::string> prefixes = {"gfx6", "gfx7", "gfx8", "gfx10"};
+  if (std::any_of(prefixes.begin(), prefixes.end(), [&](const std::string &p) {
+        return arch.find(p) != std::string::npos;
+      })) {
     return true;
-  if (arch.find("gfx7") != std::string::npos)
-    return true;
-  if (arch.find("gfx8") != std::string::npos)
-    return true;
-  if ((arch.find("gfx9") != std::string::npos) && !isCDNA(device))
-    return true;
-  if (arch.find("gfx10") != std::string::npos)
-    return true;
+  }
+
+  // gfx9 is supported if it is CDNA
+  return ((arch.find("gfx9") != std::string::npos) && !isCDNA(device));
 #else
   // NVIDIA: older than Volta is not supported
   const int arch_numeric = std::stoi(arch.substr(3));

--- a/include/ccglib/common/arch.h
+++ b/include/ccglib/common/arch.h
@@ -16,13 +16,17 @@ bool isCDNA2(cu::Device &device) {
 
 bool isCDNA3(cu::Device &device) {
   const std::string arch(device.getArch());
-  return ((arch.find("gfx940") != std::string::npos) ||
-          (arch.find("gfx941") != std::string::npos) ||
-          (arch.find("gfx942") != std::string::npos));
+  return (arch.find("gfx94") != std::string::npos);
+}
+
+bool isCDNA4(cu::Device &device) {
+  const std::string arch(device.getArch());
+  return (arch.find("gfx95") != std::string::npos);
 }
 
 bool isCDNA(cu::Device &device) {
-  return (isCDNA1(device) || isCDNA2(device) || isCDNA3(device));
+  return (isCDNA1(device) || isCDNA2(device) || isCDNA3(device) ||
+          isCDNA4(device));
 }
 
 bool isVolta(cu::Device &device) {
@@ -30,6 +34,28 @@ bool isVolta(cu::Device &device) {
   return (arch.find("sm_70") != std::string::npos);
 }
 
+bool isUnsupported(cu::Device &device) {
+  const std::string arch(device.getArch());
+  // AMD: unsupported are gfx < 8, Vega gfx9, RDNA2=gfx10
+#if defined(__HIP_PLATFORM_AMD__)
+  if (arch.find("gfx6") != std::string::npos)
+    return true;
+  if (arch.find("gfx7") != std::string::npos)
+    return true;
+  if (arch.find("gfx8") != std::string::npos)
+    return true;
+  if ((arch.find("gfx9") != std::string::npos) && !isCDNA(device))
+    return true;
+  if (arch.find("gfx10") != std::string::npos)
+    return true;
+#else
+  // NVIDIA: older than Volta is not supported
+  const int arch_numeric = std::stoi(arch.substr(3));
+  if (arch_numeric < 70)
+    return true;
+#endif
+  return false;
+}
 } // namespace ccglib
 
 #endif // ARCH_H_

--- a/src/mma/GEMM.cpp
+++ b/src/mma/GEMM.cpp
@@ -109,6 +109,9 @@ void GEMM::Impl::check_support() {
       throw std::runtime_error("Float32 input is not supported on Volta");
     }
   }
+  if (isUnsupported(device_)) {
+    throw std::runtime_error("This architecture is not supported");
+  }
 }
 
 void GEMM::Impl::compile_kernel() {

